compiler/rustc_resolve/src/lib.rs

```diff
@@ -1684,18 +1695,36 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
         false
     }
 
+    fn nearest_ambiguity_binding(&self, binding: NameBinding<'a>) -> Option<NameBinding<'a>> {
+        if binding.ambiguity.is_some() {
+            Some(binding)
+        } else {
+            match binding.kind {
+                NameBindingKind::Import { binding, .. } => self.nearest_ambiguity_binding(binding),
+                _ => None,
+            }
+        }
+    }
+
     fn record_use(&mut self, ident: Ident, used_binding: NameBinding<'a>, is_lexical_scope: bool) {
```

Comment:
```suggestion
    fn record_use_inner(&mut self, ident: Ident, used_binding: NameBinding<'a>, is_lexical_scope: bool, warn_ambiguity: bool) {
```
No need to do the recursion with `nearest_ambiguity_binding` here, `record_use` is already called recursively below.

It would be simpler to pass the right `warn_ambiguity` to it as an additional argument.
Then do `fn record_use(&mut self, ident: Ident, used_binding: NameBinding<'a>, is_lexical_scope: bool) { self.record_use_inner(ident, used_binding, is_lexical_scope, false) }` for the top level `record_use` for convenience.